src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const results = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        if (action.keysToRemove && action.keysToRemove.length > 0) {
+          for (const key of action.keysToRemove) {
+            if (currentState.hasOwnProperty(key)) {
+              delete currentState[key];
+            }
+          }
+        }
+        break;
+      default:
+        continue;
+    }
+
+    results.push({ ...currentState });
```

Comment:
You are correctly making a copy of the `currentState` before pushing it to the results array to avoid mutations to the objects within the results array. However, according to the task checklist, the name of the object copy should be more descriptive, such as `stateClone` or `newState` instead of `currentState`.